c/tskit/tables.c

```diff
@@ -5388,6 +5381,638 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+            }
+            tail->next = x;
+            self->ibd_segments_tail[pair_num] = x;
+        }
+    }
+out:
+    return ret;
+}
+
+static int TSK_WARN_UNUSED
+ibd_finder_add_ancestry(tsk_ibd_finder_t *self, tsk_id_t input_id, double left,
+    double right, tsk_id_t output_id)
+{
+    int ret = 0;
+    tsk_segment_t *tail = self->ancestor_map_tail[input_id];
+    tsk_segment_t *x = NULL;
+
+    assert(left < right);
+    if (tail == NULL) {
+        x = ibd_finder_alloc_segment(self, left, right, output_id);
+        if (x == NULL) {
+            ret = TSK_ERR_NO_MEMORY;
+            goto out;
+        }
+        self->ancestor_map_head[input_id] = x;
+        self->ancestor_map_tail[input_id] = x;
+    } else {
+        if (tail->right == left && tail->node == output_id) {
+            tail->right = right;
+        } else {
+            x = ibd_finder_alloc_segment(self, left, right, output_id);
+            if (x == NULL) {
+                ret = TSK_ERR_NO_MEMORY;
+                goto out;
+            }
+            tail->next = x;
+            self->ancestor_map_tail[input_id] = x;
+        }
+    }
+out:
+    return ret;
+}
+
+// Define functions that initialise the ibd_finder from a table collection/
+// (for now -- need number of samples, which comes from tree seq.)
+static int
+ibd_finder_init_samples(tsk_ibd_finder_t *self)
+{
+    int ret = 0;
+    size_t j;
+    tsk_id_t u;
+
+    /* Go through the sample pairs to define samples. */
+    for (j = 0; j < 2 * self->num_pairs; j++) {
+        u = self->pairs[j];
+
+        // Maybe some of these checks should happen in the Python interface instead?
+        if (u < 0 || u > (tsk_id_t) self->tables->nodes.num_rows) {
+            ret = TSK_ERR_NODE_OUT_OF_BOUNDS;
+            goto out;
+        }
+
+        if (!self->is_sample[u]) {
+            self->is_sample[u] = true;
+            ret = ibd_finder_add_ancestry(self, u, 0, self->tables->sequence_length, u);
+            if (ret != 0) {
+                goto out;
+            }
+            // Initialise the sample id map and sample list.
+            self->samples[j] = u;
+            self->sample_id_map[u] = j;
+        }
+    }
+    // for (j = 0; j < self->num_samples; j++) {
+    //     if (samples[j] < 0 || samples[j] > (tsk_id_t) self->tables->nodes.num_rows) {
+    //         ret = TSK_ERR_NODE_OUT_OF_BOUNDS;
+    //         goto out;
+    //     }
+    //     if (self->is_sample[samples[j]]) {
+    //         ret = TSK_ERR_DUPLICATE_SAMPLE;
+    //         goto out;
+    //     }
+    //     self->is_sample[samples[j]] = true;
+    //     ret = ibd_finder_add_ancestry(
+    //         self, samples[j], 0, self->tables->sequence_length, samples[j]);
+    //     if (ret != 0) {
+    //         goto out;
+    //     }
+
+    //     // Initialise the sample id map.
+    //     self->sample_id_map[samples[j]] = j;
+    // }
+out:
+    return ret;
+}
+
+static void
+ibd_finder_init_oldest_parents(tsk_ibd_finder_t *self)
```

Comment:
Are we using this now? If not, we should delete it. It's easy enough to reinstate later and having code that doesn't do anything lying around is a bad idea.